PyDSS/node_voltage_metrics.py

```diff
@@ -0,0 +1,384 @@
+
+    """
+
+    FILENAME = "voltage_metrics.json"
+
+    def __init__(self, prop, start_time, resolution, window_size):
+        self._start_time = start_time
+        self._resolution = resolution
+        self._range_a_limits = prop.limits
+        self._range_b_limits = prop.limits_b
+        self._window_size = window_size
+        self._node_names = None
+        self._metric_1_time_steps = []
+        self._metric_2_violation_counts = []
+        self._metric_3_time_steps = []
+        self._metric_4_violations = []
+        self._metric_5_min_violations = []
+        self._metric_5_max_violations = []
+        self._metric_6_time_points_outside_range_b = 0
+        self._bufs = None
+        self._num_time_points = 0
+
+    def _create_summary(self, metric_1, metric_2, metric_3, metric_5, metric_6):
+        moving_window_minutes = int(
+            (self._window_size * self._resolution).total_seconds() / 60
+        )
+        max_pnvdoaa = max((x.duration for x in metric_2.values())).total_seconds()
+        vdbaab = metric_1.duration.total_seconds()
+        mmavdoa = metric_3.duration.total_seconds()
+        max_voltage_overall = max(metric_5.max_voltages.values())
+        min_voltage_overall = min(metric_5.min_voltages.values())
+
+        num_nodes_always_inside_range_a = 0
+        num_nodes_any_outside_range_a_no_b = 0
+        num_nodes_any_outside_range_b = 0
+        for node in self._node_names:
+            min_voltage = metric_5.min_voltages[node]
+            max_voltage = metric_5.max_voltages[node]
+            if (
+                min_voltage < self._range_b_limits.min
+                or max_voltage > self._range_b_limits.max
+            ):
+                num_nodes_any_outside_range_b += 1
+            elif (
+                min_voltage < self._range_a_limits.min
+                or max_voltage > self._range_a_limits.max
+            ):
+                num_nodes_any_outside_range_a_no_b += 1
+            else:
+                num_nodes_always_inside_range_a += 1
+
+        return VoltageMetricsSummaryModel(
+            voltage_duration_between_ansi_a_and_b_minutes=vdbaab / 60,
+            max_per_node_voltage_duration_outside_ansi_a_minutes=max_pnvdoaa / 60,
+            moving_average_voltage_duration_outside_ansi_a_minutes=mmavdoa / 60,
+            moving_window_minutes=moving_window_minutes,
+            max_voltage=max_voltage_overall,
+            min_voltage=min_voltage_overall,
+            num_nodes_always_inside_ansi_a=num_nodes_always_inside_range_a,
+            num_nodes_any_outside_ansi_a_always_inside_ansi_b=num_nodes_any_outside_range_a_no_b,
+            num_nodes_any_outside_ansi_b=num_nodes_any_outside_range_b,
+            num_time_points_with_ansi_b_violations=metric_6.num_time_points,
+            total_num_time_points=self._num_time_points,
+            total_simulation_duration=self._num_time_points * self._resolution,
```

Comment:
total_simulation_duration_<minutes?|seconds?>

Reply:
It is actually a timedelta object, so I think this is a better name. It gets serialized to seconds in the JSON file. People should not be reading it there unless they know what they're doing.